api/src/services/report/validate.js

```diff
@@ -1,67 +1,37 @@
-const _ = require('underscore');
-
-/*
- *  return array of errors or empty array.
- */
-exports.validate = (def, form_data) => {
-  const missing_fields = [];
-
-  const _isDefined = obj => !_.isUndefined(obj) && !_.isNull(obj);
-
-  _.each(def.fields, (field, k) => {
-    const orig_key = k;
-    let key = k.split('.');
-    let data = form_data;
-
-    while(key.length > 1) {
-      data = data[key.shift()];
-    }
-
-    key = key[0];
+const exists = function(value) {
+  return value !== undefined && value !== null;
+};
 
-    if (field.required) {
-      if (
-        !data ||
-        !_isDefined(data[key]) ||
-        (_.isArray(data[key]) && !_isDefined(data[key][0]))
-      ) {
-        missing_fields.push(orig_key);
+exports.validate = (def, formData) => {
+  const errors = [];
+  const missingFields = [];
+
+  Object.keys(def.fields).forEach(k => {
+    const field = def.fields[k];
+    const value = formData[k];
+
+    if (field.required && !exists(value)) {
+      missingFields.push(k);
+    } else if (exists(value) && !field.list) {
+      // list is an undocumented but currently used feature of SMS parsing, where it converts
+      // integers into a translated string. Validation runs after parsing for SMS, so a field with a
+      // list parameter would fail validation.
+
+      if (field.type === 'integer' && (typeof value !== 'number' || !Number.isInteger(value))) {
+        errors.push({code: 'sys.incorrect_type', ctx: {expectedType: 'integer', key: k}});
+      } else if (field.type === 'string' && typeof value !== 'string') {
+        errors.push({code: 'sys.incorrect_type', ctx: {expectedType: 'string', key: k}});
+      } else if (field.type === 'boolean' && typeof value !== 'boolean') {
+        errors.push({code: 'sys.incorrect_type', ctx: {expectedType: 'boolean', key: k}});
+      } else if (field.type === 'complex' && typeof value !== 'object') {
+        errors.push({code: 'sys.incorrect_type', ctx: {expectedType: 'complex', key: k}});
       }
     }
   });
 
-  if (!_.isEmpty(missing_fields)) {
-    return [{code: 'sys.missing_fields', fields: missing_fields}];
-  }
-
-  if (def.validations) {
```

Comment:
According to my grep fu no one uses this validation feature (it's also not documented), which are basically a raw eval over the context of this module (so incredibly fragile against refactors etc) which can do whatever they want and require that you write a large clumsy function to implement instead of our standard `expr` pattern, e.g.: `function() { if (!form_data.blah) { return 'some error text'; }}`